api/layer/layer.go

```diff
@@ -556,81 +566,128 @@ func (n *layer) CopyObject(ctx context.Context, p *CopyObjectParams) (*data.Exte
 	})
 }
 
-func getRandomOID() (oid.ID, error) {
-	b := [32]byte{}
-	if _, err := rand.Read(b[:]); err != nil {
-		return oid.ID{}, err
-	}
-
-	var objID oid.ID
-	objID.SetSHA256(b)
-	return objID, nil
-}
-
 func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, settings *data.BucketSettings, obj *VersionedObject) *VersionedObject {
-	if len(obj.VersionID) != 0 || settings.Unversioned() {
-		var nodeVersion *data.NodeVersion
-		if nodeVersion, obj.Error = n.getNodeVersionToDelete(ctx, bkt, obj); obj.Error != nil {
-			return dismissNotFoundError(obj)
-		}
-
-		if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
-			if strings.Contains(obj.Error.Error(), "2050 message = object is locked") {
-				return obj
+	if settings.VersioningEnabled() {
+		if len(obj.VersionID) > 0 {
+			var deleteOID oid.ID
+
+			if obj.VersionID == data.UnversionedObjectVersionID {
+				versions, err := n.searchAllVersionsInNeoFS(ctx, bkt, bkt.Owner, obj.Name, true)
+				if err != nil {
+					obj.Error = fmt.Errorf("search versions: %w", err)
+					if errors.Is(err, ErrNodeNotFound) {
+						obj.Error = nil
+					}
+
+					return obj
+				}
+
+				if len(versions) == 0 {
+					obj.Error = nil
+					return obj
+				}
+
+				for _, version := range versions {
+					if obj.Error = n.objectDelete(ctx, bkt, version.GetID()); obj.Error != nil {
+						return obj
+					}
+				}
+			} else {
+				if err := deleteOID.DecodeString(obj.VersionID); err != nil {
+					obj.Error = fmt.Errorf("decode version: %w", err)
+					return obj
+				}
+
+				if obj.Error = n.objectDelete(ctx, bkt, deleteOID); obj.Error != nil {
```

Comment:
The answer seems to be here. And tombstones are to be deleted after expiration anyway. Maybe not a problem.